### PR TITLE
Add user storage guide

### LINKED ISSUE #135: Document object storage and scratch bucket
### Context

Our hubs are in the cloud, which means that a powerful and scalable way to interact with data is via cloud object storage. The Pangeo-style hubs are (afaik) all configured with an object storage "scratch bucket", meaning that object storage is part of 2i2c's service.

Our new LEAP and M2LInES hubs will need user-facing documentation that explains how to work with object storage. This documentation can live inside or outside 2i2c. Currently 2i2c docs do not address object storage at all. **I believe it would be advantageous to us all if this were part of 2i2c documentation**.

### Proposal

I propose we start by migrating some of the content in

https://pangeo.io/cloud.html#cloud-object-storage

to 

https://docs.2i2c.org/en/latest/admin/howto/data.html

Overall, a longer and more detailed "data management" section, without more information about how to work with data in the cloud, should be our goal.

I can commit to writing this documentation if we agree it is in-scope for 2i2c.

### Updates and actions

_No response_

### PR DESCRIPTION
**Pull request by rabernat**
- Closes #135 
- Starts to address #129 by adding a new section to the docs called "Hub User Guide", for which the storage page is currently the only entry.

### CODE REVIEW

File: user/storage.md
Comment (rabernat) [inline user/storage.md:39]: I know that @yuvipanda will not approve of the security practices recommended in this section. I am happy to replace this with something else that is more secure. We just need to give a clear answer to the question _how should I push to a github remote from 2i2c hubs_?

File: user/storage.md
Comment (rabernat) [inline user/storage.md:153]: Can we think of some more concrete suggestions to put here?

File: user/storage.md
Comment (yuvipanda) [reply user/storage.md:39]: @fperez has been dogfooding https://github.com/yuvipanda/github-app-user-auth for a few months now and is pretty happy with it. https://github.com/yuvipanda/github-app-user-auth/pull/8 also makes it work inline from within JupyterLab. 

I agree we should provide a clear answer to 'how do I push to github from a hub'. I can replace this section with something using https://github.com/yuvipanda/github-app-user-auth.

File: user/storage.md
Comment (choldgraf) [inline user/storage.md:169]:
```suggestion
appropriate permissions to read / write to your private bucket.
```

File: user/storage.md
Comment (choldgraf) [inline user/storage.md:170]:
```suggestion
You upload your service account key (a `.json` file) to your hub
```

File: user/storage.md
Comment (choldgraf) [inline user/storage.md:187]:
```suggestion
bucket are pre-loaded into your Hub's user environment.
```

File: user/storage.md
Comment (choldgraf) [inline user/storage.md:186]:
```suggestion
allows you to temporarily store data (for example, when you need to store intermediate files during data transformations).
Credentials to write to the scratch
```

File: user/storage.md
Comment (choldgraf) [inline user/storage.md:194]: This is probably something other than `PANGEO_SCRATCH` - what is it?

File: user/storage.md
Comment (choldgraf) [inline user/storage.md:196]:
```suggestion
For example, here is how you would write Xarray data to the scratch bucket
```

File: user/storage.md
Comment (choldgraf) [inline user/storage.md:203]: Should update these when we confirm the name of the environment variable

**Review by choldgraf — COMMENTED** [review]
I really like this - I'm always impressed at the clarity and usefulness of the Pangeo docs. Thanks so much @rabernat for porting this over.

I made several comments and suggestions throughout to improve the clarity and readability. There are also a few places where I think we might need to update the names of things (in particular the `PANGEO_SCRATCH` variable name...but correct me if I'm wrong). I also added a big "warning" to the instructions about GitHub credentials storage, so that we can merge this PR without blocking on the longer-term fix that @yuvipanda mentions.

IMO once the comments are resolved, we can merge this one in.

File: user/storage.md
Comment (damianavila) [reply user/storage.md:39]: > I can replace this section with something using https://github.com/yuvipanda/github-app-user-auth.

@yuvipanda, do you want to do that as part of this PR or maybe as a follow-up as @choldgraf suggested below?

**Comment by damianavila** [comment]
@rabernat, Chris left a bunch of comments, can I ask for a follow-up from you since you were the original author of this PR?
Thanks!!

**Comment by rabernat** [comment]
Thanks for the feedback everyone! I have a busy M-W but will try to find time to revise this on Thursday.

File: user/storage.md
Comment (yuvipanda) [reply user/storage.md:39]: I've spent a bunch of time upgrading https://github.com/yuvipanda/gh-scoped-creds over the last couple days, I'll make that a PR to this PR.

File: user/storage.md
Comment (rabernat) [reply user/storage.md:194]: On our 2i2c-managed clusters, I believe it is still PANGEO_SCRATCH. But I'd be happy to see it renamed.

Comment (yuvipanda) [reply user/storage.md:194]: SCRATCH_BUCKET and PANGEO_SCRATCH are both set. I think we should recommend SCRATCH_BUCKET

File: user/storage.md
Comment (yuvipanda) [reply user/storage.md:39]: yeah, it'll be one per hub. So user workflow would be:

1. run 'gh-scoped-creds' (or `%ghscopedcreds`) from within the notebook
2. Grant this particular JupyterHub access
3. Manage which repos this particular repo has push access to.

The initial setup for hub admins isn't too big either, so I think this just becomes a 'feature' you can ask for when we make a hub.

**Comment by yuvipanda** [comment]
I'll try to get in my gh-scoped-creds documentation PR to this PR before thursday then :)

File: user/storage.md
Comment (yuvipanda) [reply user/storage.md:39]: https://user-images.githubusercontent.com/30430/164171109-22a18151-a031-46b5-8e6f-7d0bc28cddc6.mp4

I made a quick demo of how this will look like. We can improve this over time of course. It lets users provide per-hub, per-repo permissions in a time-scoped manner. The second part where you're choosing the repo need not be done every 8h.

Comment (yuvipanda) [reply user/storage.md:39]: I turned this into an announcement blog post https://blog.jupyter.org/securely-pushing-to-github-from-a-jupyterhub-3ee42dfdc54f that is a draft in the Jupyter blog. There's yet another big supply chain attack going on right now based on stolen GitHub credentials, so good timing for this :D

File: user/storage.md
Comment (yuvipanda) [reply user/storage.md:42]: There's an active attack involving stolen github credentials going on right now (https://github.blog/2022-04-15-security-alert-stolen-oauth-user-tokens/) :)

File: user/storage.md
Comment (yuvipanda) [reply user/storage.md:39]: I'm working on setting this up on the pangeo GCP hub. https://github.com/pangeo-data/pangeo-docker-images/pull/305 is currently waiting for review.

Comment (yuvipanda) [reply user/storage.md:39]: I've test-deployed this onto https://staging.us-central1-b.gcp.pangeo.io - you can test it out by opening a notebook there, then:

```
%pip install gh-scoped-creds
import gh_scoped_creds
%ghscopedcreds
```

And following the prompts there. The `pip install` will go away once https://github.com/pangeo-data/pangeo-docker-images/pull/305 is merged.

https://github.com/2i2c-org/infrastructure/pull/1219 sets this config up and also adds docs on how to do it, so we can turn this on as a 'feature' for any new hub.

Comment (yuvipanda) [reply user/storage.md:39]: And here's my PR to this PR replacing the ssh-key section! https://github.com/rabernat/2i2c-docs/pull/1

**Comment by yuvipanda** [comment]
https://github.com/rabernat/2i2c-docs/pull/1 is my PR to this PR :)

File: user/storage.md
Comment (rabernat) [inline user/storage.md:185]:
```suggestion
The location of your scratch bucket is contained in the environment variable ``SCRATCH_BUCKET ``.
```

File: user/storage.md
Comment (rabernat) [inline user/storage.md:30]: Is it true that you can't ssh / scp into your server? Or do we have a solution for that? If so, this would be a good place to mention it.

File: user/storage.md
Comment (rabernat) [reply user/storage.md:39]: This is an amazing solution.

File: user/storage.md
Comment (rabernat) [inline user/storage.md:111]: The link checker is erroneously flagging this as a broken link

```
(    user/storage: line  109) broken    https://www.openstoragenetwork.org/ - HTTPSConnectionPool(host='www.openstoragenetwork.org', port=443): Max retries exceeded with url: / (Caused by SSLError(SSLCertVerificationError(1, '[SSL: CERTIFICATE_VERIFY_FAILED] certificate verify failed: unable to get local issuer certificate (_ssl.c:1131)')))
```

Comment (choldgraf) [inline user/storage.md:111]:
```suggestion
  services (e.g. [Wasabi](https://wasabi.com/) and [Open Storage Newtork](https://openstoragenetwork.org/))
```

Comment (choldgraf) [reply user/storage.md:111]: this might fix the URL error

Comment (choldgraf) [reply user/storage.md]: I added an exception to openstoragenetwork, not sure why it wasn't working

File: conf.py
Comment (choldgraf) [inline conf.py:69]:
```suggestion
    "https://openstoragenetwork.org*",  # It incorrectly fails with `Max retries exceeded with url`
```

File: user/storage.md
Comment (yuvipanda) [reply user/storage.md:30]: https://github.com/yuvipanda/jupyterhub-ssh adds SSH and SFTP (but not scp) support. Maybe we could make that an additional feature we can deploy. I know that @consideRatio has this going for @fperez's JMTE hub for a while - how has your experience been, @consideRatio?

Comment (consideRatio) [reply user/storage.md:30]: I think we've had a few surprises learning about some limitations, but have been happy about what is supported.

- `rsync` over SSH -> SFTP recommended to use instead -> Some confusion about that we had to use another port than port 22 for that as we had to distinguish the JupyterHub SSH server from the JupyterHub SFTP server.
  https://github.com/yuvipanda/jupyterhub-ssh/issues/55
- Some challenges about servers being culled by `jupyterhub-idle-culler` when having an open SSH connection
  https://github.com/yuvipanda/jupyterhub-ssh/issues/67
- SSH Tunnels aren't supported, but perhaps an expectation held for someone installing `jupyterhub-ssh`
  https://github.com/yuvipanda/jupyterhub-ssh/issues/26

Comment (rabernat) [reply user/storage.md:30]: Let's leave ssh out of this guide for now. We can always update it later.

I also think it's fine if we try to push people towards a more cloud-native approach to data.

**Review by yuvipanda — APPROVED** [review]
This is good to go!

**Comment by yuvipanda** [comment]
Thanks, @rabernat and @choldgraf!

**Pull request merged**
